portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,178 @@
-@WebServlet("/data")
+@WebServlet("/comments")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    
+    String queryType = request.getParameter("location");
+    System.err.println("queryType: " + queryType);
+    Query query = new Query(queryType);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Comment> messages = new ArrayList<Comment>();
+
+    int numIterations = 0;
+
+    for (Entity entity : results.asIterable()) {
+      if (numIterations >= globalNumComments) {
+          break;
+      }
+      numIterations++;
+
+      String name = (String) entity.getProperty("name");
+      String message = (String) entity.getProperty("message");
+
+      Comment newComment = new Comment(name, message);
+      
+      messages.add(newComment);
+    }
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(messages));
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    System.out.println("Add new message to Comments section");
+
+    String name = request.getParameter("commenter-name");
+    String newComment = request.getParameter("text-input");
+    String entityType = request.getParameter("location");
+    System.out.println("entityType: " + entityType);
+
+    Location loc = Location.valueOf(entityType);
+
+    Entity taskEntity = new Entity(entityType);
+    taskEntity.setProperty("name", name);
+    taskEntity.setProperty("message", newComment);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(taskEntity);
+
+    globalNumComments = getNumComments(request);
+    System.out.println("globalNumComments: " + globalNumComments);
+
+    // Redirect back to the HTML page, using Location enum
+    redirectPage(loc, response);
+
+  }
+
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    String location = request.getParameter("location");
+    Location loc = Location.valueOf(location);
+    System.out.println("PUT location: " + location);
+
+    globalNumComments = getNumComments(request);
+    System.out.println("globalNumComments: " + globalNumComments);
+
+    // Redirect back to the HTML page, using Location enum
+    redirectPage(loc, response);
+  }
+
+  /** Returns the choice entered by the player, or -1 if the choice was invalid. */
+  private int getNumComments(HttpServletRequest request) {
+
+    String numCommentsString = request.getParameter("num-comments");
+
+    // Convert the input to an int.
+    int numComments = 10;
```

Comment:
We don't want random numbers like 10 (very confusing where this number comes from). Instead, make that a constant like default_num_comments = 10

Then on L115, do this
int numComments = default_num_comments